pkg/apis/pipeline/v1beta1/param_types.go

```diff
@@ -110,6 +111,205 @@ type Param struct {
 	Value ParamValue `json:"value"`
 }
 
+// extractParamValuesFromParams get all param values from params
+func extractParamValuesFromParams(params []Param) []string {
```

Comment:
this can be defined as a struct function or a member function:

```
type Params []Param

// extractParamValuesFromParams get all param values from params
func (ps Params) extractParamValuesFromParams() []string {
	pvs := []string{}
	for i := range ps {
		pvs = append(pvs, ps[i].Value.StringVal)
		pvs = append(pvs, ps[i].Value.ArrayVal...)
		for _, v := range ps[i].Value.ObjectVal {
			pvs = append(pvs, v)
		}
	}
	return pvs
}
```

https://github.com/tektoncd/pipeline/commit/5c3869083f6183188cb43deedd39aedbfead6236